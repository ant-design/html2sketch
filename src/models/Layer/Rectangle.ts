import { BaseLayerParams, SketchFormat } from '../../types';

import BaseLayer from '../Base/BaseLayer';
import { defaultExportOptions } from '../utils';

type CornerRadius = {
  bottomLeft: number;
  bottomRight: number;
  topLeft: number;
  topRight: number;
};
interface RectangleInitParams extends Omit<BaseLayerParams, 'x' | 'y'> {
  x?: number;
  y?: number;
  cornerRadius?: CornerRadius | number | number[];
}
/**
 * 矩形类型
 * */
class Rectangle extends BaseLayer {
  constructor({
    x,
    y,
    width,
    height,
    cornerRadius = { topLeft: 0, bottomLeft: 0, topRight: 0, bottomRight: 0 },
  }: RectangleInitParams) {
    super(SketchFormat.ClassValue.Rectangle, { height, x, y, width });

    this.cornerRadius = cornerRadius;
  }

  /**
   * 圆角值
   */
  cornerRadius: CornerRadius | number | number[] = 0;

  toKonvaRadius = () => {
    if (
      typeof this.cornerRadius === 'number' ||
      this.cornerRadius instanceof Array
    ) {
      return this.cornerRadius;
    }
    return Object.values(this.cornerRadius);
  };

  /**
   * 转换为 Sketch JSON
   */
  toSketchJSON(): SketchFormat.Rectangle {
    return {
      _class: 'rectangle',
      name: this.name,
      resizingConstraint: this.resizingConstraint,
      frame: this.frame.toSketchJSON(),
      do_objectID: this.id,
      hasConvertedToNewRoundCorners: true,
      needsConvertionToNewRoundCorners: false,
      fixedRadius: 0,
      style: this.style.toSketchJSON(),
      edited: false,
      pointRadiusBehaviour: 1,
      points: this.getSketchPoints(),
      isClosed: true,
      booleanOperation: SketchFormat.BooleanOperation.NA,
      exportOptions: defaultExportOptions,
      isVisible: true,
      isFixedToViewport: false,
      isFlippedHorizontal: false,
      isFlippedVertical: false,
      layerListExpandedType: 0,
      nameIsFixed: false,
      resizingType: 0,
      rotation: this.rotation,
      shouldBreakMaskChain: false,
      clippingMaskMode: 0,
      isLocked: false,
      hasClippingMask: this.hasClippingMask,
    };
  }

  /**
   * 转换为 Konva JSON
   */
  toKonvaJSON() {
    const { cornerRadius } = this;
    return {
      attrs: {
        ...this.frame.toJSON(),
        id: this.id,
        cornerRadius:
          typeof cornerRadius === 'number' || cornerRadius instanceof Array
            ? cornerRadius
            : [
                cornerRadius.topLeft,
                cornerRadius.topRight,
                cornerRadius.bottomRight,
                cornerRadius.bottomLeft,
              ],
      },
      className: this.name,
    };
  }

  /**
   * 获取 SketchPoints
   */
  getSketchPoints = (): SketchFormat.CurvePoint[] => {
    const { cornerRadius } = this;
    let topRight;
    let topLeft;
    let bottomLeft;
    let bottomRight;
    if (typeof cornerRadius === 'number') {
      topLeft = cornerRadius;
      topRight = cornerRadius;
      bottomRight = cornerRadius;
      bottomLeft = cornerRadius;
    } else if (cornerRadius instanceof Array) {
      [topLeft, topRight, bottomRight, bottomLeft] = cornerRadius;
    } else {
      topLeft = cornerRadius.topLeft;
      topRight = cornerRadius.topRight;
      bottomRight = cornerRadius.bottomRight;
      bottomLeft = cornerRadius.bottomLeft;
    }
    return [
      {
        _class: 'curvePoint',
        cornerRadius: topLeft,
        curveFrom: '{0, 0}',
        curveMode: 1,
        curveTo: '{0, 0}',
        hasCurveFrom: false,
        hasCurveTo: false,
        point: '{0, 0}',
      },
      {
        _class: 'curvePoint',
        cornerRadius: topRight,
        curveFrom: '{1, 0}',
        curveMode: 1,
        curveTo: '{1, 0}',
        hasCurveFrom: false,
        hasCurveTo: false,
        point: '{1, 0}',
      },
      {
        _class: 'curvePoint',
        cornerRadius: bottomRight,
        curveFrom: '{1, 1}',
        curveMode: 1,
        curveTo: '{1, 1}',
        hasCurveFrom: false,
        hasCurveTo: false,
        point: '{1, 1}',
      },
      {
        _class: 'curvePoint',
        cornerRadius: bottomLeft,
        curveFrom: '{0, 1}',
        curveMode: 1,
        curveTo: '{0, 1}',
        hasCurveFrom: false,
        hasCurveTo: false,
        point: '{0, 1}',
      },
    ];
  };
}

export default Rectangle;
