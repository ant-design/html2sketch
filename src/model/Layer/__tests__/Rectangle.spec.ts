import Rectangle from '../Rectangle';

describe('Rectangle 类', () => {
  describe('toKonvaJSON 转换正常', () => {
    it('普通 Rectangle 转换正常', () => {
      const rect = new Rectangle({ height: 100, width: 50, x: 280, y: 100 });
      rect.name = 'Rect';
      expect(rect.toKonvaJSON()).toStrictEqual({
        attrs: {
          cornerRadius: [0, 0, 0, 0],
          x: 280,
          y: 100,
          width: 50,
          id: 'UUID',
          height: 100,
        },
        className: 'Rect',
      });
    });
    it('等圆角 Rectangle 转换正常', () => {
      const rect = new Rectangle({
        height: 100,
        width: 50,
        x: 280,
        y: 100,
        cornerRadius: 10,
      });
      rect.name = 'Rect';
      expect(rect.toKonvaJSON()).toStrictEqual({
        attrs: {
          x: 280,
          y: 100,
          width: 50,
          id: 'UUID',
          height: 100,
          cornerRadius: 10,
        },
        className: 'Rect',
      });
    });
    it('不等圆角 Rectangle 转换正常', () => {
      const rect = new Rectangle({
        height: 100,
        width: 50,
        x: 280,
        y: 100,
        cornerRadius: {
          bottomLeft: 10,
          bottomRight: 2,
          topLeft: 3,
          topRight: 2,
        },
      });
      rect.name = 'Rect';
      expect(rect.toKonvaJSON()).toStrictEqual({
        attrs: {
          x: 280,
          y: 100,
          width: 50,
          id: 'UUID',
          height: 100,
          cornerRadius: [3, 2, 2, 10],
        },
        className: 'Rect',
      });
    });
    it('不等圆角 Rectangle2 转换正常', () => {
      const rect = new Rectangle({
        height: 100,
        width: 50,
        x: 280,
        y: 100,
        cornerRadius: [10, 10, 2, 4],
      });
      rect.name = 'Rect';
      expect(rect.toKonvaJSON()).toStrictEqual({
        attrs: {
          x: 280,
          y: 100,
          width: 50,
          id: 'UUID',
          height: 100,
          cornerRadius: [10, 10, 2, 4],
        },
        className: 'Rect',
      });
    });
  });
  describe('toSketchJSON 转换正常', () => {
    it('正常转换', () => {
      const rect = new Rectangle({ height: 100, width: 50, x: 280, y: 100 });
      rect.name = 'Rect';
      expect(rect.toSketchJSON()).toMatchSnapshot();
      expect(rect.toSketchJSON()).toStrictEqual({
        _class: 'rectangle',
        booleanOperation: -1,
        clippingMaskMode: 0,
        do_objectID: 'UUID',
        edited: false,
        exportOptions: {
          _class: 'exportOptions',
          exportFormats: [],
          includedLayerIds: [],
          layerOptions: 0,
          shouldTrim: false,
        },
        fixedRadius: 0,
        frame: {
          _class: 'rect',
          constrainProportions: false,
          height: 100,
          width: 50,
          x: 280,
          y: 100,
        },
        hasConvertedToNewRoundCorners: true,
        isClosed: true,
        isFixedToViewport: false,
        isFlippedHorizontal: false,
        isFlippedVertical: false,
        isLocked: false,
        isVisible: true,
        layerListExpandedType: 0,
        name: 'Rect',
        nameIsFixed: false,
        needsConvertionToNewRoundCorners: false,
        pointRadiusBehaviour: 1,
        points: [
          {
            _class: 'curvePoint',
            cornerRadius: 0,
            curveFrom: '{0, 0}',
            curveMode: 1,
            curveTo: '{0, 0}',
            hasCurveFrom: false,
            hasCurveTo: false,
            point: '{0, 0}',
          },
          {
            _class: 'curvePoint',
            cornerRadius: 0,
            curveFrom: '{1, 0}',
            curveMode: 1,
            curveTo: '{1, 0}',
            hasCurveFrom: false,
            hasCurveTo: false,
            point: '{1, 0}',
          },
          {
            _class: 'curvePoint',
            cornerRadius: 0,
            curveFrom: '{1, 1}',
            curveMode: 1,
            curveTo: '{1, 1}',
            hasCurveFrom: false,
            hasCurveTo: false,
            point: '{1, 1}',
          },
          {
            _class: 'curvePoint',
            cornerRadius: 0,
            curveFrom: '{0, 1}',
            curveMode: 1,
            curveTo: '{0, 1}',
            hasCurveFrom: false,
            hasCurveTo: false,
            point: '{0, 1}',
          },
        ],
        resizingConstraint: 63,
        resizingType: 0,
        rotation: 0,
        shouldBreakMaskChain: false,
        style: {
          _class: 'style',
          borderOptions: {
            _class: 'borderOptions',
            dashPattern: [],
            isEnabled: true,
            lineCapStyle: 0,
            lineJoinStyle: 0,
          },
          borders: [],
          colorControls: {
            _class: 'colorControls',
            brightness: 0,
            contrast: 1,
            hue: 0,
            isEnabled: false,
            saturation: 1,
          },
          contextSettings: {
            _class: 'graphicsContextSettings',
            blendMode: 0,
            opacity: 1,
          },
          do_objectID: 'UUID',
          endMarkerType: 0,
          fills: [],
          innerShadows: [],
          miterLimit: 10,
          shadows: [],
          startMarkerType: 0,
          windingRule: 1,
        },
      });
    });
  });
});
