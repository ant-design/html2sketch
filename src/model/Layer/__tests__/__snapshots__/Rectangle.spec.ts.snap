// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rectangle 类 toSketchJSON 转换正常 正常转换 1`] = `
Object {
  "_class": "rectangle",
  "booleanOperation": -1,
  "clippingMaskMode": 0,
  "do_objectID": "UUID",
  "edited": false,
  "exportOptions": Object {
    "_class": "exportOptions",
    "exportFormats": Array [],
    "includedLayerIds": Array [],
    "layerOptions": 0,
    "shouldTrim": false,
  },
  "fixedRadius": 0,
  "frame": Object {
    "_class": "rect",
    "constrainProportions": false,
    "height": 100,
    "width": 50,
    "x": 280,
    "y": 100,
  },
  "hasConvertedToNewRoundCorners": true,
  "isClosed": true,
  "isFixedToViewport": false,
  "isFlippedHorizontal": false,
  "isFlippedVertical": false,
  "isLocked": false,
  "isVisible": true,
  "layerListExpandedType": 0,
  "name": "Rect",
  "nameIsFixed": false,
  "needsConvertionToNewRoundCorners": false,
  "pointRadiusBehaviour": 1,
  "points": Array [
    Object {
      "_class": "curvePoint",
      "cornerRadius": 0,
      "curveFrom": "{0, 0}",
      "curveMode": 1,
      "curveTo": "{0, 0}",
      "hasCurveFrom": false,
      "hasCurveTo": false,
      "point": "{0, 0}",
    },
    Object {
      "_class": "curvePoint",
      "cornerRadius": 0,
      "curveFrom": "{1, 0}",
      "curveMode": 1,
      "curveTo": "{1, 0}",
      "hasCurveFrom": false,
      "hasCurveTo": false,
      "point": "{1, 0}",
    },
    Object {
      "_class": "curvePoint",
      "cornerRadius": 0,
      "curveFrom": "{1, 1}",
      "curveMode": 1,
      "curveTo": "{1, 1}",
      "hasCurveFrom": false,
      "hasCurveTo": false,
      "point": "{1, 1}",
    },
    Object {
      "_class": "curvePoint",
      "cornerRadius": 0,
      "curveFrom": "{0, 1}",
      "curveMode": 1,
      "curveTo": "{0, 1}",
      "hasCurveFrom": false,
      "hasCurveTo": false,
      "point": "{0, 1}",
    },
  ],
  "resizingConstraint": 63,
  "resizingType": 0,
  "rotation": 0,
  "shouldBreakMaskChain": false,
  "style": Object {
    "_class": "style",
    "borderOptions": Object {
      "_class": "borderOptions",
      "dashPattern": Array [],
      "isEnabled": true,
      "lineCapStyle": 0,
      "lineJoinStyle": 0,
    },
    "borders": Array [],
    "colorControls": Object {
      "_class": "colorControls",
      "brightness": 0,
      "contrast": 1,
      "hue": 0,
      "isEnabled": false,
      "saturation": 1,
    },
    "contextSettings": Object {
      "_class": "graphicsContextSettings",
      "blendMode": 0,
      "opacity": 1,
    },
    "do_objectID": "UUID",
    "endMarkerType": 0,
    "fills": Array [],
    "innerShadows": Array [],
    "miterLimit": 10,
    "shadows": Array [],
    "startMarkerType": 0,
    "windingRule": 1,
  },
}
`;
